### PR TITLE
Library call guard.

### PR DESCRIPTION
**Pull request by chriseth**
Still needs to be properly checked.

### CODE REVIEW

File: test/cmdlineTests/standard_ewasm_requested/output.json.orig [hunk 1]
```diff
@@ -0,0 +1,171 @@
+<<<<<<< HEAD
```
Comment (hrkrshnn) [inline test/cmdlineTests/standard_ewasm_requested/output.json.orig]: a leftover file?

Comment (aarlt) [reply test/cmdlineTests/standard_ewasm_requested/output.json.orig]: hmm.. maybe we should add a check in style check that ensures that there are no ".orig" files

Comment (aarlt) [reply test/cmdlineTests/standard_ewasm_requested/output.json.orig]: or `.gitignore`

File: libyul/backends/wasm/polyfill/Memory.yul [hunk 2]
```diff
@@ -61,6 +61,15 @@ function memoryguard(x:i64) -> y1, y2, y3, y4 {
 	y4 := x
 }
 
+function linkersymbol(x) -> y1, y2, y3, y4 {
+}
```
Comment (axic) [inline libyul/backends/wasm/polyfill/Memory.yul]: I would put an `invalid()` into these. And likely wouldn't put them into `Memory.yul`, if there's a better place. However not including them here is also good, we just won't enable these tests?

Comment (aarlt) [reply libyul/backends/wasm/polyfill/Memory.yul]: Turned out that the additional polyfills are not really needed.

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 3]
```diff
@@ -806,7 +822,18 @@ string IRGenerator::dispatchRoutine(ContractDefinition const& _contract)
...
+		{
+			solAssert(!type->isPayable(), "");
+			if (type->stateMutability() > StateMutability::View)
+				// If the function is not a view function and is called without DELEGATECALL,
+				// we revert.
+				// TODO add revert message.
+				delegatecallCheck = "if iszero(called_via_delegatecall) { revert(0, 0) }";
+		}
+		templ["delegatecallCheck"] = delegatecallCheck;
+		templ["callValueCheck"] = (type->isPayable() || _contract.isLibrary()) ? "" : callValueCheck();
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGenerator.cpp:836]: Just added the `|| _contract.isLibrary()` today.

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 4]
```diff
@@ -806,7 +822,18 @@ string IRGenerator::dispatchRoutine(ContractDefinition const& _contract)
...
 		templ["functionName"] = type->externalSignature();
-		templ["callValueCheck"] = type->isPayable() ? "" : callValueCheck();
+		string delegatecallCheck;
+		if (_contract.isLibrary())
+		{
+			solAssert(!type->isPayable(), "");
+			if (type->stateMutability() > StateMutability::View)
+				// If the function is not a view function and is called without DELEGATECALL,
+				// we revert.
+				// TODO add revert message.
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGenerator.cpp:832]: This 'TODO' is present for all revert messages in sol->yul.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 5]
```diff
@@ -3554,52 +3581,55 @@ BOOST_AUTO_TEST_CASE(mapping_returns_in_library)
...
-	ABI_CHECK(callContractFunction("get(bool,uint256)", true, u256(21)), encodeArgs(u256(14)));
-	ABI_CHECK(callContractFunction("get_b(uint256)", u256(0)), encodeArgs(u256(0)));
-	ABI_CHECK(callContractFunction("get_b(uint256)", u256(1)), encodeArgs(u256(30)));
-	ABI_CHECK(callContractFunction("get_b(uint256)", u256(2)), encodeArgs(u256(31)));
-	ABI_CHECK(callContractFunction("get_b(uint256)", u256(21)), encodeArgs(u256(32)));
-	ABI_CHECK(callContractFunction("get(bool,uint256)", false, u256(0)), encodeArgs(u256(0)));
-	ABI_CHECK(callContractFunction("get(bool,uint256)", false, u256(1)), encodeArgs(u256(30)));
-	ABI_CHECK(callContractFunction("get(bool,uint256)", false, u256(2)), encodeArgs(u256(31)));
-	ABI_CHECK(callContractFunction("get(bool,uint256)", false, u256(21)), encodeArgs(u256(32)));
+	ALSO_VIA_YUL(
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp:3584]: It would be nice if some helpful angel could extract some of these...

**Review by ekpyron — APPROVED** [review]

**Pull request merged**